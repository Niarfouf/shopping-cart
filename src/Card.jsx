export default function Card({ title, price, description, imageUrl }) {
    return (
        <>
            <div className="card">
                <img className="item-image" alt={title} src={imageUrl}></img>
                <h3 className="title">{title}</h3>
                <p>{description}</p>
                <input type="number"></input>
                <button>Add</button>
                <p>{price}</p>
            </div>
        </>
    )
}
