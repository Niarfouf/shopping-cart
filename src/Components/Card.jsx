import { useEffect, useState } from 'react'
export default function Card({ item, addItemInList, quantity = 0, action }) {
    const [input, setInput] = useState(0)
    function left() {
        setInput((input) => (input > 0 ? input - 1 : input))
    }
    function right() {
        setInput((input) => input + 1)
    }
    useEffect(() => {
        setInput(quantity)
    }, [quantity])
    return (
        <>
            <div className="card">
                <img
                    className="item-image"
                    alt={item.title}
                    src={item.image}
                ></img>
                <h3 className="title">{item.title}</h3>

                <div>
                    <div className="add-container">
                        <label htmlFor={item.id}>Quantity :</label>
                        <button onClick={left} className="left-arrow-input">
                            <svg
                                xmlns="http://www.w3.org/2000/svg"
                                viewBox="0 0 24 24"
                            >
                                <title>menu-left</title>
                                <path d="M14,7L9,12L14,17V7Z" />
                            </svg>
                        </button>
                        <input
                            value={input}
                            onChange={(e) => setInput(e.target.value * 1)}
                            type="text"
                            inputMode="numeric"
                            name="quantity"
                            id={item.id}
                        ></input>
                        <button onClick={right} className="right-arrow-input">
                            <svg
                                xmlns="http://www.w3.org/2000/svg"
                                viewBox="0 0 24 24"
                            >
                                <title>menu-right</title>
                                <path d="M10,17L15,12L10,7V17Z" />
                            </svg>
                        </button>
                        <button
                            className="add-button"
                            onClick={() => addItemInList(item, input)}
                        >
                            {action}
                        </button>
                    </div>
                    <p className="price">
                        {input} unit x {item.price}€ ={' '}
                        {Math.round(item.price * input * 100) / 100} €
                    </p>
                </div>
            </div>
        </>
    )
}
