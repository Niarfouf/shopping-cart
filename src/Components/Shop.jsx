import { useParams } from 'react-router-dom'

import { useState, useEffect } from 'react'
import Card from './Card'

export default function Shop({ addCartItems }) {
    const { category } = useParams()
    const [items, setItems] = useState([])
    useEffect(() => {
        async function fetchItems() {
            try {
                const response = await fetch(
                    `https://fakestoreapi.com/products/category/${category}?limit=8`
                )
                const data = await response.json()
                setItems(data)
            } catch (e) {
                console.error(e)
            }
        }
        fetchItems()
    }, [category])
    function capitalizeFirstLetter(string) {
        return string.charAt(0).toUpperCase() + string.slice(1)
    }
    return (
        <>
            <div className="shop">
                <h2>{capitalizeFirstLetter(category)}</h2>
                <div className="item-cards">
                    {items.map((item) => {
                        return (
                            <Card
                                key={item.id}
                                item={item}
                                addItemInList={addCartItems}
                                action={'Add'}
                            ></Card>
                        )
                    })}
                </div>
            </div>
        </>
    )
}
