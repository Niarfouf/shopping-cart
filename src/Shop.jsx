import { useParams } from 'react-router-dom'
import { useState, useEffect } from 'react'
import Card from './Card'

export default function Shop() {
    const { category } = useParams()
    const [items, setItems] = useState([])
    useEffect(() => {
        async function fetchItems() {
            try {
                const response = await fetch(
                    `https://fakestoreapi.com/products/category/${category}?limit=8`
                )
                const data = await response.json()
                setItems(data)
            } catch (e) {
                console.error(e)
            }
        }
        fetchItems()
    }, [category])

    return (
        <>
            <h1>test shop</h1>
            <div className="item-cards">
                {items.map((item) => {
                    return (
                        <Card
                            key={item.id}
                            title={item.title}
                            price={item.price}
                            description={item.description}
                            imageUrl={item.image}
                        ></Card>
                    )
                })}
            </div>
        </>
    )
}
