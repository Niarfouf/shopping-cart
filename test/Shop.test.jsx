import { vi, beforeAll, afterAll } from 'vitest'
import { render, act, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Shop from '../src/Components/Shop'
import Card from '../src/Components/Card'
import { MemoryRouter, Route, Routes } from 'react-router-dom'

describe('Nav', () => {
    //Spy on the global fetch function
    const fetchSpy = vi.spyOn(window, 'fetch')

    //Run before all the tests
    beforeAll(() => {
        //Mock the return value of the global fetch function
        const mockResolveValue = {
            ok: true,
            json: () =>
                new Promise((resolve) =>
                    resolve([
                        {
                            title: 'testTitle1',
                            image: 'https://fakestoreapi.com/img/61U7T1koQqL._AC_SX679_.jpg',
                            id: 0,
                            price: 5.5,
                        },
                        {
                            title: 'testTitle2',
                            image: 'https://fakestoreapi.com/img/71kWymZ+c+L._AC_SX679_.jpg',
                            id: 1,
                            price: 10,
                        },
                    ])
                ),
        }

        fetchSpy.mockReturnValue(mockResolveValue)
    })
    afterAll(() => {
        fetchSpy.mockRestore()
    })
    it('render snapshot', async () => {
        const category = 'electronics'
        const { container } = render(
            <MemoryRouter initialEntries={[`/shop/${category}`]}>
                <Routes>
                    <Route
                        path="/shop/:category"
                        element={<Shop addCartItems={() => {}} />}
                    ></Route>
                </Routes>
            </MemoryRouter>
        )

        await waitFor(() => expect(fetchSpy).toHaveBeenCalledTimes(1))

        expect(container).toMatchSnapshot()
    })
    it('render correct number of products', async () => {
        const category = 'electronics'
        render(
            <MemoryRouter initialEntries={[`/shop/${category}`]}>
                <Routes>
                    <Route
                        path="/shop/:category"
                        element={<Shop addCartItems={() => {}} />}
                    ></Route>
                </Routes>
            </MemoryRouter>
        )

        await waitFor(() => expect(fetchSpy).toHaveBeenCalledTimes(2))

        const heading = screen.getAllByRole('heading')
        expect(heading.length).toBe(3)
        expect(heading[0].textContent).toBe('Electronics')
    })
})
