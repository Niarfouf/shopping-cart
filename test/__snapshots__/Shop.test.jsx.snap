// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Nav > render snapshot 1`] = `
<div>
  <div
    class="shop"
  >
    <h2>
      Electronics
    </h2>
    <div
      class="item-cards"
    >
      <div
        class="card"
      >
        <img
          alt="testTitle1"
          class="item-image"
          src="https://fakestoreapi.com/img/61U7T1koQqL._AC_SX679_.jpg"
        />
        <h3
          class="title"
        >
          testTitle1
        </h3>
        <div>
          <div
            class="add-container"
          >
            <label
              for="0"
            >
              Quantity :
            </label>
            <button
              class="left-arrow-input"
            >
              <svg
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <title>
                  menu-left
                </title>
                <path
                  d="M14,7L9,12L14,17V7Z"
                />
              </svg>
            </button>
            <input
              id="0"
              inputmode="numeric"
              name="quantity"
              type="text"
              value="0"
            />
            <button
              class="right-arrow-input"
            >
              <svg
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <title>
                  menu-right
                </title>
                <path
                  d="M10,17L15,12L10,7V17Z"
                />
              </svg>
            </button>
            <button
              class="add-button"
            >
              Add
            </button>
          </div>
          <p
            class="price"
          >
            0
             unit x 
            5.5
            € =
             
            0
             €
          </p>
        </div>
      </div>
      <div
        class="card"
      >
        <img
          alt="testTitle2"
          class="item-image"
          src="https://fakestoreapi.com/img/71kWymZ+c+L._AC_SX679_.jpg"
        />
        <h3
          class="title"
        >
          testTitle2
        </h3>
        <div>
          <div
            class="add-container"
          >
            <label
              for="1"
            >
              Quantity :
            </label>
            <button
              class="left-arrow-input"
            >
              <svg
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <title>
                  menu-left
                </title>
                <path
                  d="M14,7L9,12L14,17V7Z"
                />
              </svg>
            </button>
            <input
              id="1"
              inputmode="numeric"
              name="quantity"
              type="text"
              value="0"
            />
            <button
              class="right-arrow-input"
            >
              <svg
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <title>
                  menu-right
                </title>
                <path
                  d="M10,17L15,12L10,7V17Z"
                />
              </svg>
            </button>
            <button
              class="add-button"
            >
              Add
            </button>
          </div>
          <p
            class="price"
          >
            0
             unit x 
            10
            € =
             
            0
             €
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
